### PR TITLE
AO3-5039 Prevent non canonical tags from being favorited

### PR DESCRIPTION
**Pull request by sarken**
## Issue

https://otwarchive.atlassian.net/browse/AO3-5039

## Purpose

Makes it so trying to favorite a non canonical tag gives an error message. Also adds tests for the favorite tags model and fixes an issue found when writing those specs.

## Testing

Refer to JIRA issue

### CODE REVIEW

File: app/models/favorite_tag.rb [hunk 1]
```diff
@@ -11,16 +11,24 @@ class FavoriteTag < ActiveRecord::Base
...
 
   validate :within_limit, on: :create
-
-  after_save :expire_cached_home_favorite_tags
-  after_destroy :expire_cached_home_favorite_tags
-
   def within_limit
-    if user.favorite_tags(:reload).count >= ArchiveConfig.MAX_FAVORITE_TAGS
-      errors.add(:base, ts('Sorry, you can only save %{maximum} favorite tags.', maximum: ArchiveConfig.MAX_FAVORITE_TAGS))
+    if user && user.favorite_tags(:reload).count >= ArchiveConfig.MAX_FAVORITE_TAGS
```
Comment (sarken) [inline app/models/favorite_tag.rb]: Without `user` in here, the new spec for this failed with:
```
  2) FavoriteTag is invalid without a user_id
     Failure/Error: if user.favorite_tags(:reload).count >= ArchiveConfig.MAX_FAVORITE_TAGS
     
     NoMethodError:
       undefined method `favorite_tags' for nil:NilClass
     # ./app/models/favorite_tag.rb:15:in `within_limit'
     # ./spec/models/favorite_tag_spec.rb:25:in `block (2 levels) in <top (required)>'
```

File: app/models/favorite_tag.rb [hunk 2]
```diff
@@ -11,16 +11,24 @@ class FavoriteTag < ActiveRecord::Base
...
+      errors.add(:base,
+                 ts("Sorry, you can only save %{maximum} favorite tags.",
+                    maximum: ArchiveConfig.MAX_FAVORITE_TAGS))
+    end
+  end
+
+  validate :canonical, on: :create
```
Comment (zz9pzza) [inline app/models/favorite_tag.rb]: What would happen if someone favorited a tag and it was changed to non canonical ?

Comment (sarken) [inline app/models/favorite_tag.rb]: We have code in place that destroys a favorite tag when it's de-canonized. That code shouldn't conflict with this -- it's currently tested here, and all tests are passing: https://github.com/otwcode/otwarchive/blob/master/features/tags_and_wrangling/favorite_tags.feature#L31

**Comment by sarken** [comment]
This had some test failures, possibly due to changes in the tag factory since this was originally written. I've fixed them and this needs to be re-reviewed.

File: spec/models/favorite_tag_spec.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    create(:favorite_tag, tag_id: tag.id, user_id: user.id)
+    expect(build(:favorite_tag, tag_id: tag.id, user_id: user.id).save).to be_falsey
+  end
+
+  it "is invalid without a user_id" do
+    expect(no_user_favorite.save).to be_falsey
+  end
+
+  it "is invalid with a non-canonical tag" do
+    expect(noncanonical_favorite.save).to be_falsey
```
Comment (redsummernight) [inline spec/models/favorite_tag_spec.rb]: Some of these `save` calls can be replaced with `valid?`. We can test validations without saving to the database which makes it faster.

**Pull request merged**